docs/proposals/20250325-distributed-inference.md

```diff
@@ -133,35 +132,135 @@ env:
+| **Worker** | vLLM does not gracefully handle worker pod restarts (see [vLLM #16259](https://github.com/vllm-project/vllm/issues/16259)), as a restarted worker joins the existing Ray cluster but remains idle, rendering the inference service unusable despite the Ray cluster appearing healthy. That said, implementing a liveness probe for worker pods is unnecessary, as worker health is indirectly validated through the leader’s liveness probe. The leader’s probe, by detecting dead actors, triggers a complete Ray cluster reinitialization, synchronizing both leader and worker pods to restore service functionality. | Checks leader’s `/health` endpoint at `$VLLM_LEADER_SVC:5000` after some initial delay. |
+
+
+The following sequential diagram summarizes the expected behavior of the liveness and readiness probes during different failure scenarios:
+
+#### Leader Pod Crash
+
+```mermaid
+sequenceDiagram
+    participant K as Liveness Probes
+    participant P as Readiness Probes
+    participant L as Leader Pod
+    participant W as Worker Pod
+    participant R as Ray Cluster
+    participant V as Inference Service
+
+    Note over L,W: Initialization Phase: Ray cluster creation, workers joining leader
+
+    %% Leader Crash During Initialization
+    L->>W: Waits for all workers to join
+    W->>L: Attempting to join Ray cluster
+    L--xL: Leader crashes
+    Note over V: Inference service unavailable (no leader)
+    K->>L: Liveness probe checks dead Ray actors
+    K-->>L: Probe fails, triggers leader container restart
+    L->>R: Restarts and recreates Ray cluster
+    W->>L: Reconnects to new leader (1–2 min delay)
+    Note over W: May heartbeat to old leader
+    L->>W: Waits for all workers to join
+    W->>L: All workers join
+    Note over V: Inference service becomes available
+
+    %% Leader Crash After Inference Service is Ready
+    V->>L: Serving inference requests
+    L--xL: Leader crashes
+    Note over V: Inference service unavailable (no leader)
+    P->>L: Readiness probe check
+    P-->>L: Probe fails, marks pod not ready
+    K->>L: Liveness probe checks dead Ray actors
+    K-->>L: Probe fails, triggers leader container restart
+    L->>R: Restarts and recreates Ray cluster
+    W->>L: Reconnects to new leader (1–2 min delay)
+    Note over W: May heartbeat to old leader
+    L->>W: Waits for all workers to join
+    W->>L: All workers join
+    Note over V: Inference service becomes available again
+```
+
+#### Worker Pod Crash
+
+```mermaid
```

Comment:
Wow nice diagrams. So based on this multinode is completely fault tolerant?

Reply:
In theory, yes. A mechanism to restart the leader container on error should be sufficient to cover most failure scenarios, but there might be other unknown scenarios I didn’t account for, which I’d need to identify during testing.